### DISCUSSION TITLE
python-oracledb thin client artificial packet during login

### DISCUSSION THREAD

**Discussion by explorercpp** [oracle/python-oracledb#184] [Q&A] [answered]
When recording tcpdump using python-oracledb thin client, during connection process I can see an artificial network packet of size 1 with value 0x21.

`
Client -> Server
00000000 : 21                                                                     !

Client -> Server
00000000 : 00 00 00 0a 0c 20 00 00 02 00                   ..... ....

Client <- Server
00000000 : 00 00 00 0a 0e 20 00 00 00 09                   ..... ....
`

Can someone please explain what is this single byte packet?
I tried same python-oracledb library with instant client and such network packet is not present. Also tried sqlplus, was not present.
What could such packet sent from client to server mean? Does it indicate something special?
It was packet 18 during the login process.

Thank you in advance for any thoughts.

**Comment by anthony-tuininga** [comment, marked as answer]
I believe this is the OOB (out-of-band) detection that is available with certain servers. You can turn it off by passing the parameter `disable_oob` as the value `True` when creating the connection. SQL*Plus (and python-oracledb thick mode) depend on the equivalent value found in the `sqlnet.ora` configuration file.

**Reply by explorercpp** [reply]
Thank you! Found this parameter in the connect_params.py of the thin library.
`disable_oob: bool=False`
Will try to disable it and see if this packet goes away.

**Reply by anthony-tuininga** [reply]
You're welcome! Was that packet causing you problems? It is solely intended to determine if OOB processing is possible between the client and the server. If you know that OOB processing is not possible then disabling it makes sense, of course!

**Reply by explorercpp** [reply]
I'm trying to understand if there will be different network load for different connection types from different clients and authentication methods, and sometimes artificial content appears which I never saw before, such as OOB packet. And your input is very helpful. Especially after I switched to new Mac with ARM processor and many things stopped working as expected for unknown reasons

**Reply by explorercpp** [reply]
Is there some existing reference of such OOB codes? I'm pretty sure that with variety of systems, value 0x21 is not the only one that might exist. Thank you again

**Reply by explorercpp** [reply]
changed it to 
```
./connection.py:1043:            disable_oob: bool=True,
./connect_params.py:80:                 disable_oob: bool=True,
```

Still 0x21 packet appears in tcpdump. I didn't change the pool.py, but I'm not using pooling connections. Tried on 2 different laptops, seems same result.

Would you think it might be something else?

**Reply by anthony-tuininga** [reply]
If you are using *thick* mode you have to set the value in your `sqlnet.ora` file; otherwise it will have no effect. For thin mode you don't need to change any code, you can simply do this:

```
import oracledb
conn = oracledb.connect(user=USER, password=PASSWORD, dsn=DSN, disable_oob=True)
```

I confirmed with my own database that the 0x21 byte is indeed the "urgent" byte sent (it is the exclamation mark (!) symbol) as part of the test.

**Reply by explorercpp** [reply]
Indeed, code change did not affect it, but specifying it in the connection string did remove that packet. Seems there are other parameters to play with in thin mode, will explore them. Thank you

**Comment by explorercpp** [comment]
If I may may ask further questions about network packets.

I was looking at the connections, and I can see that the username that was used to login is places in this packet:

<img width="1582" alt="Screenshot 2023-05-10 at 4 25 30 PM" src="https://github.com/oracle/python-oracledb/assets/60533393/dbd87d75-1b7c-4ad0-aea3-0ef530bf3782">

Above is example of login made by user scott.

But in some cases, I can see in this place value "1". And such user does not exist. 

<img width="1025" alt="Screenshot 2023-05-10 at 4 26 22 PM" src="https://github.com/oracle/python-oracledb/assets/60533393/155180c6-d2d7-49b1-808d-20ceb57e1b1b">


I was able to reproduce such case, when I tried something like "sqlplus "1"/as sysdba". But none of my applications are doing such connection.

If you have any idea what this value "1" means, will be very helpful.

**Reply by anthony-tuininga** [reply]
Those are packets using thick mode and I am not as familiar with how those work. Can you reproduce with thin mode?

**Reply by explorercpp** [reply]
Most of my apps are using thick mode, in thin mode I have only one app so far on python-oracledb. I was trying to reproduce it, but haven't seen it yet. May I ask if there is such discussion channel for thick mode? Maybe worth to give it a try, if your colleagues might know thick area

**Reply by anthony-tuininga** [reply]
I can point someone who knows thick mode to this question and see if they can answer. :-)

**Reply by explorercpp** [reply]
Thank you so much! You are the man of the day :)

**Reply by anthony-tuininga** [reply]
I got an answer that indicates that is indeed the username. You can verify that yourself by doing something like this:

```
create user "1" identified by some_password;
grant create session to "1";
```

At that point you should be able to connect and see the packet dump matches.

**Reply by explorercpp** [reply]
Exactly, that's another way to reproduce this packet. But in my case, this user does not exist, and some connections already show such packet. Could it be some script, service, management or back up job that makes a local connection to Oracle with internal user "1"? That connection does not fail, even user does not exist.

**Reply by anthony-tuininga** [reply]
I've asked and will let you know if I get an answer. I don't know myself!

**Reply by explorercpp** [reply]
Sure, appreciate your time and help in passing the info to your colleagues!

**Reply by anthony-tuininga** [reply]
You're welcome. As for the answer: I am informed that there is no way for that to happen without someone explicitly doing it -- so you will need to find out from your end why that is happening!

**Reply by explorercpp** [reply]
Ok, I will try to narrow it down and identify specifically which connection and at what time creates such connection. Thank you!
